src/Shared/PackageActions/NuGetPackageActions.cs

```diff
@@ -0,0 +1,169 @@
+            packageUrl.Name,
+            NuGetVersion.Parse(packageUrl.Version),
+            packageStream,
+            _sourceCacheContext,
+            _logger, 
+            cancellationToken);
+        
+        // If the download failed, return null.
+        if (!downloaded)
+        {
+            return null;
+        }
+        
+        // If we want to extract the contents of the .nupkg, send it to ArchiveHelper.ExtractArchiveAsync.
+        if (doExtract)
+        {
+            return await ArchiveHelper.ExtractArchiveAsync(topLevelDirectory, Path.GetExtension(targetPath) + ".nupkg", packageStream, cached);
+        }
+
+        return filePath;
+    }
+
+    /// <inheritdoc />
+    public async Task<bool> DoesPackageExistAsync(
+        PackageURL packageUrl,
+        bool useCache = true,
+        CancellationToken cancellationToken = default)
+    {
+        FindPackageByIdResource resource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
+
+        // If a version is provided, check for existence of the specified version.
+        if (!string.IsNullOrWhiteSpace(packageUrl.Version))
+        {
+            bool exists = await resource.DoesPackageExistAsync(
+                packageUrl.Name,
+                NuGetVersion.Parse(packageUrl.Version),
+                _sourceCacheContext,
+                _logger, 
+                cancellationToken);
+
+            return exists;
+        }
+        
+        // If no version is provided, check to see if any versions exist on a package with the given name.
+        IEnumerable<NuGetVersion> versions = await resource.GetAllVersionsAsync(
+            packageUrl.Name,
+            _sourceCacheContext,
+            _logger, 
+            cancellationToken);
+
+        return versions.Any();
+    }
+
+    /// <inheritdoc />
+    public async Task<IEnumerable<string>> GetAllVersionsAsync(
+        PackageURL packageUrl,
+        bool includePrerelease = true,
+        bool useCache = true,
+        CancellationToken cancellationToken = default)
+    {
+        FindPackageByIdResource resource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
+
+        IEnumerable<NuGetVersion> versions = await resource.GetAllVersionsAsync(
+            packageUrl.Name,
+            _sourceCacheContext,
+            _logger, 
+            cancellationToken);
+
+        return versions
+            .Where(v => includePrerelease || !v.IsPrerelease)
+            .Select(v => v.ToString()).Reverse();
+    }
+    
+    /// <inheritdoc />
+    public async Task<string?> GetLatestVersionAsync(
+        PackageURL packageUrl,
+        bool includePrerelease = false,
+        bool useCache = true,
+        CancellationToken cancellationToken = default)
+    {
+        FindPackageByIdResource resource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
+
+        IEnumerable<NuGetVersion> versions = await resource.GetAllVersionsAsync(
```

Comment:
Same thought as above - what do you think about including the sort order as part of the name for this variable? Perhaps something like `versionsOldestFirst`?

Reply:
`versionsAscending`?